In [1]:
#Reproduce the analysis of the Hard to Employ programs for NY.
#Reproduce the results in cell 2 of Table 2.1 (Ever employed in a CEO 
#transitional job), and cell 10 (Convicted of a felony). Fill in the 
#cells of the scheleton notebook as you are asked to.

#The second null hypothesis should be stated (for the "Convicted of a 
#felony after 3 years" data).

#Both tests, Z and chi-sq, should be completed for the "Convicted of a 
#felony after 3 years" data.

#The result of the test in term the rejection of the Null should be 
#stated in all cases (for both tests and both for the original "Ever 
#employed in a CEO transitional job" data and the "Convicted of a felony 
#after 3 years data").

In [6]:
import os
import sys
import numpy as np
import pylab as pl
import pandas as pd
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [6]:
#null hypothesis: the percent of former prisoners employed after release
#is the same or lower for candidates who participated in the program as
#for the control group. 
#p=.05
#H0: P0 (control) - P1 (program) >= 0
#HA: P0 (control) - P1 (program) < 0

In [7]:
#conducting a test of proportions, which is a binomial distribution question
#for each participant, the former inmate either was or was not ever 
#employed in a CEO transitional job

In [8]:
#P_0 = 3.5 * .01
#P_1 = 70.1 * .01
#n_0 = 409
#n_1 = 564

In [9]:
#given this information, we have two options. we can use a z test or a chi
#squared test.
#the z test compares the standard deviation of the expected distribution 
#and the observed result. it tells you how many standard deviations from 
#the tail an observation is, under the assumption of normality

In [10]:
#define the sample proportion first
#sp = (number of people in control group employed in CEO transitional job)
#+ (number of people in program group employed in CEO transitional job) /
#(total sample size, control group plus program group)

#sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
#print (sp)

In [11]:
#the standard deviation equation for a binomial distribution is:
#square_root((p(1 - p))/n)

#the standard deviation approximation for two samples with a binomial
#distribution is:
#square root((p(1-p)/n_0) + ((p(1-p))/n_1))
#estimate 'p' with the following equation:
#(P_0*n_0+P_1*n_1)/(n_0+n_1)

In [12]:
#can use the below formula to compute a z statistic for the diference in
#proportions between two populations
#sample size assumption is (np >= 5, n(1-p) >=5) (for both samples in this
#case)
#remember, any reference to 'proportion' is a reference to the .035 and .701
#values that were calculated from the percentage of people in the control
#group and in the program group who ever had a CEO transitional job

In [13]:
#the z statistic is calculated with the following equation:
#(proportion 1 - proportion 2) / sample standard deviation
#in this case, proportions 1 and 2 are control (P0) and program (P1)
#respectively

In [14]:
#to calculate the standard error, AKA the estimate of the sampling 
#distribution standard deviation, do the following:
#calculate the estimate of p, which is (P_0 * n_0 + P_1 * n_1) / (n_0 + n_1)
#then use p in the standard error formula:
#se = np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1)) 

In [15]:
#next, we calculate the z score, which answers the question of how many 
#standard deviations away from the population parameter:
#z score = P_0 - P_1 / pop_SD

In [16]:
#WHAT IS THIS?
#z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))

In [17]:
#if the z statistic (OR Z SCORE?) is less than the significance value 
#stated above, then reject the null hypothesis

In [18]:
#p-value = 1-P #WHAT IS THIS?
#WHAT IS THIS? p_2y = 1 - 0.9998

### Question 1: 
Was there a difference in recidivism rates of the control and program groups for those individuals who were convicted of a felony?

#### Null Hypothesis:
The recidivism rate of those individuals who were convicted of felony crimes is lower for those who participated in the 'Hard-to-Employ' program than for those who did not participate in the program. 

#### H0: 
prop_control - prop_program <= 0

#### Alternative Hypothesis: 
The recidivism rate of those individuals who were convicted of felony crimes is lower for those who participated in the 'Hard-to-Employ' program than for those who did not participate in the program. 

#### HA:
prop_control - prop_program > 0

#### Significance Level: 
p = .05

In [1]:
prop_control = 11.7 * .01 #P0
prop_program = 10.0 * .01 #P1
n_control = 409 #n0
n_program = 564 #n1
alpha = .05

In [2]:
if prop_control - prop_program <= 0:
    # we are done
    print("the Null holds")
else:
    print("must assess the statistical significance")

must assess the statistical significance


In [3]:
#define the sample proportion first
sample_sd = (prop_control * n_control + prop_program * n_program) / (n_control + n_program)
print(sample_sd)

0.107145940391


In [4]:
#pooled sample proportion
p = lambda prop_control, prop_program, n_control, n_program: (prop_control * n_control + prop_program * n_program) / (n_control + n_program)

#standard error
se = lambda p, n_control, n_program: np.sqrt(p * (1 - p) * (1.0 / n_control + 1.0 / n_program)) 

In [8]:
zscore = lambda prop_control, prop_program, s : (prop_control - prop_program) / s
z_2y = zscore(prop_control, prop_program, se(p(prop_control, prop_program, n_control, n_program), n_control, n_program))
print(z_2y)

0.846282982605


In [10]:
#because the z score is .846, it's a good idea to calculate the average table value by taking the z score of .8 
#and .9
area = (.8023+.8289)/2
p_2y = 1 - area

def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")  
    print ('the null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )
report_result(p_2y, alpha)

is the p value 0.18 smaller than the critical value 0.05?
NO!
the null hypothesis is not rejected


In [37]:
#how to read the chi-squared table is two-dimensional because the distribution
#changes based on the number of degrees of freedom

#degrees of freedom = # of observations - # dependent variables
#for a test of proportions between two samples, DF = 1

#go to the chi-squared table and check the expected value based on the
#degrees of freedom and p-value. if the chi-squared statistic is 
#greater than the minimum chi squared significance value, then you can
#reject the null hypothesis

#the chi squared statistic is calculated by summing the squared distances
#between the observed and the expected values and dividing by the 
#expected value.
#chi_stat = ((control_count-expected)**2 + (program_count-expected)**2)/expected

### Question 2:

Null and alternative hypotheses and significance level remain the same for the chi-squared test as they were for the z test.

In [59]:
sample_values = np.array([[0.117 * 409, 0.883 * 409], [0.1 * 564, 0.9 * 564]])
print(sample_values)

[[  47.853  361.147]
 [  56.4    507.6  ]]


In [56]:
expected = sample_values[0][0] + sample_values[1][0]
print("DOF = 1")
chi_stat = ((control_count-expected)**2 + (program_count-expected)**2)/expected
print(chi_stat)

DOF = 1
52.4768554286


In [58]:
def evalChisq(values):
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()
print(evalChisq(sample_values))

0.716194886647


Because the chi-squared score was .716 and the chi-squared distribution table indicated the value of .05 significance and 1 DOF is equal to 3.84, I canot reject the null hypothesis.